### PR TITLE
fix: data collection for headless router 

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?
- Headless Router(/router?form={FORM_ID}) was just doing the routing but not collecting the form responses. This PR fixes that
- Comma separated strings were supported for Routing Forms prefill and Headless router but they weren't working correctly. It's fixed now.

[Demo showing Headless router collects data now](https://www.loom.com/share/6cf0056e1fb840b883c5997bb56c3bf4)

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How should this be tested?

- [x] Create a Routing Form with different types of fields and see that reporting shows the collected data
- [x] Create a Routing Form with different types of fields and see that routing is working with all types of fields.

## Checklist
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by hariombalhara** [comment]
It's an important PR. Will work on it one high priority things are done

File: packages/app-store/routing-forms/trpc/utils.ts [hunk 1]
```diff
@@ -58,8 +57,11 @@ export const sendResponseEmail = async (
   ownerEmail: string
 ) => {
   try {
-    const email = new ResponseEmail({ form: form, toAddresses: [ownerEmail], response: response });
-    await email.sendEmail();
+    if (typeof window === "undefined") {
+      const ResponseEmail = (await import("../emails/templates/response-email")).default;
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/trpc/utils.ts:61]: Even though this function isn't included browser side, I don't know why it tries to be built for client side code and thus throwing the error.

So, I have to do it.

File: packages/app-store/routing-forms/lib/transformResponse.ts [hunk 2]
```diff
@@ -8,5 +8,20 @@ export default function transformResponse({
...
+    if (typeof value === "string") {
+      return Number(value);
+    }
+    return value;
+  }
+  if (field.type === "multiselect") {
+    if (value instanceof Array) {
+      return value;
+    }
+    return value
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/lib/transformResponse.ts:21]: Handle comma separated values as well for multiselect

File: packages/app-store/routing-forms/pages/router/[...appPages].tsx [hunk 3]
```diff
@@ -34,14 +37,15 @@ const querySchema = z
     slug: z.string(),
     pages: z.array(z.string()),
   })
-  .catchall(z.string());
+  .catchall(z.string().or(z.array(z.string())));
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/pages/router/[...appPages].tsx:40]: Support comma separated string as well as multiple params with same name for multiple select.

File: packages/app-store/routing-forms/pages/router/[...appPages].tsx [hunk 4]
```diff
@@ -54,11 +58,7 @@ export const getServerSideProps = async function getServerSideProps(
     where: {
       id: formId,
       user: {
-        organization: isValidOrgDomain
-          ? {
-              slug: currentOrgDomain,
-            }
-          : null,
+        organization: isValidOrgDomain && currentOrgDomain ? getSlugOrRequestedSlug(currentOrgDomain) : null,
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/pages/router/[...appPages].tsx:61]: `getSlugOrRequestedSlug` supports unpublished org as well.

File: packages/app-store/routing-forms/pages/router/[...appPages].tsx [hunk 5]
```diff
@@ -86,6 +86,18 @@ export const getServerSideProps = async function getServerSideProps(
     throw new Error("No matching route could be found");
   }
 
+  const { createContext } = await import("@calcom/trpc/server/createContext");
+  const ctx = await createContext(context);
+
+  const trpcRouter = (await import("@calcom/app-store/routing-forms/trpc/_router")).default;
+  const caller = trpcRouter.createCaller(ctx);
+  const { v4: uuidv4 } = await import("uuid");
+  await caller.public.response({
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/pages/router/[...appPages].tsx:95]: Start collecting responses from headless router.

File: packages/app-store/routing-forms/pages/routing-link/[...appPages].tsx [hunk 6]
```diff
@@ -302,9 +303,11 @@ const usePrefilledResponse = (form: Props["form"]) => {
   form.fields?.forEach((field) => {
     const valuesFromQuery = searchParams?.getAll(getFieldIdentifier(field)).filter(Boolean);
     // We only want to keep arrays if the field is a multi-select
+
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/pages/routing-link/[...appPages].tsx:306]:
```suggestion
```

File: packages/app-store/routing-forms/pages/routing-link/[...appPages].tsx [hunk 7]
```diff
@@ -302,9 +303,11 @@ const usePrefilledResponse = (form: Props["form"]) => {
...
     const value = valuesFromQuery.length > 1 ? valuesFromQuery : valuesFromQuery[0];
+
     prefillResponse[field.id] = {
-      value: value || "",
+      value: transformResponse({ field, value }),
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/pages/routing-link/[...appPages].tsx:310]: Now, routing link also correctly handles, comma separated strings. Earlier, they were prefilled correctly but routing didn't work correctly for comma separated strings

File: packages/app-store/routing-forms/lib/transformResponse.ts [hunk 8]
```diff
@@ -8,5 +8,20 @@ export default function transformResponse({
...
+  if (field.type === "number") {
+    if (typeof value === "string") {
+      return Number(value);
+    }
+    return value;
+  }
+  if (field.type === "multiselect") {
+    if (value instanceof Array) {
+      return value;
+    }
```
Comment (ologbonowiwi) [inline packages/app-store/routing-forms/lib/transformResponse.ts:20]: Why do we have this condition if both cases are returning the value? 🤔

File: packages/app-store/routing-forms/pages/router/[...appPages].tsx [hunk 9]
```diff
@@ -86,6 +86,18 @@ export const getServerSideProps = async function getServerSideProps(
     throw new Error("No matching route could be found");
   }
 
+  const { createContext } = await import("@calcom/trpc/server/createContext");
+  const ctx = await createContext(context);
+
+  const trpcRouter = (await import("@calcom/app-store/routing-forms/trpc/_router")).default;
```
Comment (ologbonowiwi) [inline packages/app-store/routing-forms/pages/router/[...appPages].tsx:92]:
```suggestion
  const { default: trpcRouter } = await import("@calcom/app-store/routing-forms/trpc/_router");
```

File: packages/app-store/routing-forms/trpc/utils.ts (hunk 1 above)
Comment (ologbonowiwi) [inline packages/app-store/routing-forms/trpc/utils.ts:61]:
```suggestion
      const { default: ResponseEmail } = await import("../emails/templates/response-email");
```

File: packages/app-store/routing-forms/lib/transformResponse.ts (hunk 8 above)
Comment (hariombalhara) [reply packages/app-store/routing-forms/lib/transformResponse.ts:20]: It's easier to read for me. Like a switch statement there is a check for type of field and then you handle it's cases.

**Comment by hariombalhara** [comment]
Thanks for the review @ologbonowiwi. Accepted your suggestions 🙏

**Review by CarinaWolli — APPROVED** [review]
Works and code looks good 👏🏻

**Pull request merged**